### PR TITLE
Made rollout and parameter exploration sampling reproducible.

### PR DESCRIPTION
**Pull request by fdamken**
* This is done by allowing the seed in `rollout` to be a `str` which is converted to an `int` with an MD5 hash.
* The seed string is dynamically created from the number of samples that have been drawn before and the parameters passed to `sample`, namely the `init_states` and `domain_params`.
* This dynamic seeding is needed because even different number of workers should result in the same rollouts.
* Added three tests to test that the sampling is deterministic (conditioned on the seed).

Here is a TensorBoard screenshot of three runs, two with the same seed (5) and different number of workers (4 (orange) and 8 (blue)), and one with another seed (12345) (red):
image

Another screenshot, ran on commit f69d138:
image

### CODE REVIEW

**Comment by fdamken** [comment]
It might be faster and more stable to restrict the base seed to a size of 15 Bits and use the remaining 16 Bits for the sub-seed. This would allow for 65,536 different seeds for a given base seed and for 32,768 base seeds. The seeds can simply be concatenated with `(base_seed << 16) | sub_seed`.

This is less flexible than using a string seed, but more predictable.

**Comment by famura** [comment]
Thanks @fdamken for this nice contribution. I have a few questions (assuming you do it the way described in your 2nd message).
1. What happens is the user specifies an integer for the seed that is bigger than what could be represented with 16 bits?
2. Does `sub_seed` need a default value? If so, 0 would make intuitive sense to me.
3. Is it possible to remove the dependency on `init_state` and `domain_param`? That would be desirable.
4. What is `num` referring to?

I think that the more interpretable solution (2nd) is the better choice, since one typically uses only a few integer values for the seed anyways.

File: Pyrado/pyrado/sampling/parameter_exploration_sampler.py [hunk 1]
```diff
@@ -126,9 +127,9 @@ def _pes_init(G, env, policy):
     G.policy = pickle.loads(policy)
 
 
-def _pes_sample_one(G, param):
+def _pes_sample_one(G, param, seed: str):
     """Sample one rollout with the current setting."""
-    pol_param, dom_param, init_state = param
+    num, (pol_param, dom_param, init_state) = param
```
Comment (famura) [inline Pyrado/pyrado/sampling/parameter_exploration_sampler.py:132]: wouldn't it be better to pass `num` as an arg here?

File: Pyrado/pyrado/sampling/parallel_rollout_sampler.py [hunk 2]
```diff
@@ -256,29 +250,29 @@ def sample(
                     # Run every initial state so often that we at least get min_rollouts trajectories
                     func = partial(_ps_run_one_init_state, eval=eval)
                     rep_factor = ceil(self.min_rollouts / len(init_states))
-                    arglist = rep_factor * init_states
+                    arglist = list(product(range(rep_factor), init_states))
```
Comment (famura) [inline Pyrado/pyrado/sampling/parallel_rollout_sampler.py:253]: Do the two statements yield the same? If so, why did you change it?

**Comment by fdamken** [comment]
”What happens is the user specifies an integer for the seed that is bigger than what could be represented with 16 bits?“
That'd cause an exception in the `set_seed` method. I'd test it on every call, so this should fail fast and but waste die time.

”Does `sub_seed` need a default value? If so, 0 would make intuitive sense to me.“
Yes, I'd set it to `0`.

”Is it possible to remove the dependency on `init_state` and `domain_param`? That would be desirable.“
I removed it from the seed already – or do you mean something/somewhere else?

”I noticed that not every function called by the `ParallelRolloutSampler` or the `ParallelExplorationSampler` has a `num` argument. Moreover, for `run_map()` it is called counter. Please check if that is a problem for your approach.“
What do you mean exactly?

”I think that the more interpretable solution (2nd) is the better choice, since one typically uses only a few integer values for the seed anyways.“
Yes, I think that, too. But I just noticed another problem with it, but it is fixable: there is not one sub-seed, but two: the sample number (i.e., how often things where sampled) and the rollout number. The former is needed to not some the same rollouts over and over again. This means the number of bits for each of them had to be adjusted. I suggest allocating 14 bits for the sample number, allowing 16,384 sample calls it total (after that I would just wrap around to zero again, the policy should be different enough after this many iterations to not cause much problems here (i.e., it will only result in the same init state). Then 7 bits for the rollout number, allowing 128 rollouts (I would not wrap the to zero but rather just add it up to the sample count to not prevent sampling equivalent rollouts – but in theory, this many rollouts should be plenty). And finally the remaining 10 bits for the base seed, allowing 1024 different seeds. Assuming only every 10th seed is able to actually learn something, this would give approximately 100 seeds, which should be plenty.
An alternative would be to just stick to the string and hope for no collision :D

File: Pyrado/pyrado/sampling/parallel_rollout_sampler.py (hunk 2 above)
Comment (fdamken) [reply Pyrado/pyrado/sampling/parallel_rollout_sampler.py:253]: They yield the same number of rollouts, but the changed statement includes the rollout number in a tuple. But damn, this is exactly why I included the initial state in the seed at first: assuming the initial state is irrelevant for the rollout for some environment, this would sample the same rollout multiple times. But this was also the case if the same initial state was passed twice. I'll change this to give each rollout an own number.

File: Pyrado/pyrado/sampling/parameter_exploration_sampler.py (hunk 1 above)
Comment (fdamken) [reply Pyrado/pyrado/sampling/parameter_exploration_sampler.py:132]: I'll take a closer look at the `run_map` method of this would be easier. Given my comment to your other annotation,im I have to touch this again nevertheless.

Comment (fdamken) [reply Pyrado/pyrado/sampling/parameter_exploration_sampler.py:132]: I checked again, but `run_map` only supports one argument to be passed to the subroutine.

File: Pyrado/pyrado/sampling/parallel_rollout_sampler.py (hunk 2 above)
Comment (fdamken) [reply Pyrado/pyrado/sampling/parallel_rollout_sampler.py:253]: I changed it to `enumerate(rep_factor * init_states)` to give each rollout to be sampled a unique ID. The `list(…)` is necessary because `run_map` checks the `len()` and `enumerate` has no `len()`.

**Comment by fdamken** [comment]
Correction: The NumPy seed can be 32 bits long, so I would reserve 8 bits for the sub-sub-seed.

**Comment by famura** [comment]
> Correction: The NumPy seed can be 32 bits long, so I would reserve 8 bits for the sub-sub-seed.

Since we set all seeds in the pyrado method, we should go with the shortest, right? So if the others are at max 16bit...

**Comment by fdamken** [comment]
> > Correction: The NumPy seed can be 32 bits long, so I would reserve 8 bits for the sub-sub-seed.
> 
> Since we set all seeds in the pyrado method, we should go with the shortest, right? So if the others are at max 16bit...

As war as I saw the 32 Bit of NumPy are the minimum, I just first thought it where 31 Bit.

**Comment by famura** [comment]
> That'd cause an exception in the `set_seed` method. I'd test it on every call, so this should fail fast and but waste die time.
:+1: 
 
> Yes, I'd set it to `0`.
:+1: 

> ”Is it possible to remove the dependency on `init_state` and `domain_param`? That would be desirable.“
:+1: 

> ”I think that the more interpretable solution (2nd) is the better choice, since one typically uses only a few integer values for the seed anyways.“
> Yes, I think that, too. But I just noticed another problem with it, but it is fixable: there is not one sub-seed, but two: the sample number (i.e., how often things where sampled) and the rollout number. The former is needed to not some the same rollouts over and over again. This means the number of bits for each of them had to be adjusted. I suggest allocating 14 bits for the sample number, allowing 16,384 sample calls it total (after that I would just wrap around to zero again, the policy should be different enough after this many iterations to not cause much problems here (i.e., it will only result in the same init state). Then 7 bits for the rollout number, allowing 128 rollouts (I would not wrap the to zero but rather just add it up to the sample count to not prevent sampling equivalent rollouts – but in practice, this many rollouts should be plenty). And finally the remaining 10 bits for the base seed, allowing 1024 different seeds. Assuming only every 10th seed is able to actually learn something, this would give approximately 100 seeds, which should be plenty.
> An alternative would be to just stick to the string and hope for no collision :D

I agree with your explanations (exept for the "Assuming only every 10th seed is able to actually learn something" - thats very dramatic :wink: ) and the arrangement of bits per (sub)seed. However, if the hashed stirng avoids all of this descision making, which would lead to impossible-to-find bugs in very rare events, I prefer the hashed string. What do you think?

**Comment by fdamken** [comment]
Strings avoid this decision making, but in the end the seed can only be 32 Bits, no matter how it is computed. Due to the warning messages I added, I think the “manual” method is more debug-friendly. And my assumption on only every 10th seed being learnable was an upper bound ;)

**Comment by fdamken** [comment]
> Strings avoid this decision making, but in the end the seed can only be 32 Bits, no matter how it is computed. Due to the warning messages I added, I think the “manual” method is more debug-friendly. And my assumption on only every 10th seed being learnable was an upper bound ;)

I take that point back, strings might be better. Especially at the beginning of training, lots of rollouts have to be sampled. So the number of bits would have to be shifted somehow. I'll switch it back to hashes.

**Comment by famura** [comment]
> > Strings avoid this decision making, but in the end the seed can only be 32 Bits, no matter how it is computed. Due to the warning messages I added, I think the “manual” method is more debug-friendly. And my assumption on only every 10th seed being learnable was an upper bound ;)
> 
> I take that point back, strings might be better. Especially at the beginning of training, lots of rollouts have to be sampled. So the number of bits would have to be shifted somehow. I'll switch it back to hashes.

Good point. This espcially holds for the wild early exploration.

Thank you for having such a detailed look at it :+1:

File: Pyrado/pyrado/sampling/parallel_rollout_sampler.py [hunk 3]
```diff
@@ -48,6 +48,9 @@
 from pyrado.spaces.box import InfBoxSpace
 
 
+NO_SEED_PASSED = object()
```
Comment (famura) [inline Pyrado/pyrado/sampling/parallel_rollout_sampler.py:51]: just out of curiosity: why `object()`?

File: Pyrado/pyrado/sampling/step_sequence.py [hunk 4]
```diff
@@ -343,6 +343,12 @@ def __init__(
         else:
             self._rollout_bounds = None
 
+    def __repr__(self) -> str:
+        return f"StepSequence[rewards={self.rewards}, observations={self.observations}, actions={self.actions}]"
```
Comment (famura) [inline Pyrado/pyrado/sampling/step_sequence.py:347]: Please use line breaks, of even better tabulate.tabulate() here, since __repr__ sould be human readable

File: Pyrado/pyrado/sampling/step_sequence.py [hunk 5]
```diff
@@ -343,6 +343,12 @@ def __init__(
...
+
+    def __str__(self) -> str:
+        return str(self.observations)
```
Comment (famura) [inline Pyrado/pyrado/sampling/step_sequence.py:350]: Is this used for the hash?

Comment (fdamken) [reply Pyrado/pyrado/sampling/step_sequence.py:350]: No, I used it for debugging the smoke test to check whether the sampled sequences are different.

File: Pyrado/pyrado/sampling/parallel_rollout_sampler.py (hunk 3 above)
Comment (fdamken) [reply Pyrado/pyrado/sampling/parallel_rollout_sampler.py:51]: It creates a fresh instance such that the below `is` statement ist reliable. Could be anything else, e.g. an empty dict, too. But imho using a plain object instance is less confusing.

**Review by famura — APPROVED** [review]

**Comment by famura** [comment]
Thank you very much for your great contribution @fdamken
